File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjJavaFX/Server/src/main/java/app/server/ServicesImpl.java
package app.server;

import app.model.Employee;
import app.model.Game;
import app.model.TicketPurchase;
import app.persistence.EmployeesRepository;
import app.persistence.GamesRepository;
import app.persistence.TeamsRepository;
import app.persistence.TicketPurchasesRepository;
import app.services.IObserver;
import app.services.IServices;
import app.services.MyException;
import app.utils.Hash;

import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ServicesImpl implements IServices {

    private EmployeesRepository employeesRepo;
    private GamesRepository gamesRepo;
    private TeamsRepository teamsRepo;
    private TicketPurchasesRepository ticketPurchasesRepo;
    private Map<Long, IObserver> loggedEmployees;
    private final int defaultThreadsNo = 5;

    public ServicesImpl(EmployeesRepository eRepo, GamesRepository gRepo, TeamsRepository tRepo,
                        TicketPurchasesRepository tpRepo)
    {
        employeesRepo = eRepo;
        gamesRepo = gRepo;
        teamsRepo = tRepo;
        ticketPurchasesRepo = tpRepo;
        loggedEmployees = new ConcurrentHashMap<>();
    }
    @Override
    public synchronized void logIn(String username, String password, IObserver client) throws MyException {
        Employee emp = employeesRepo.findByUsername(username);
        if (emp != null) {
            if(!emp.getHashedPassword().equals(Hash.hash(password)))
                throw new MyException("Authentication failed.");
            if(loggedEmployees.get(emp.getId()) != null)
                throw new MyException("Employee already logged in!");
            loggedEmployees.put(emp.getId(), client);
        }
        else{
            throw new MyException("Authentication failed.");
        }
    }

    @Override
    public synchronized void purchase(long gameId, int noOfTickets, String clientName) throws MyException {
        Game game = gamesRepo.findById(gameId);
        float totalPrice = noOfTickets * game.getPricePerTicket();
        TicketPurchase ticketPurchase = new TicketPurchase(game, noOfTickets, totalPrice, clientName);
        ticketPurchasesRepo.add(ticketPurchase);
        game.setSoldTickets(game.getSoldTickets() + ticketPurchase.getNoOfTickets());
        gamesRepo.update(game.getId(), game);
        notifyEmployees(gamesRepo.findById(game.getId()));
    }

    @Override
    public synchronized void logOut(Employee employee, IObserver client) throws MyException {
        Employee emp = employeesRepo.findByUsername(employee.getUsername());
        IObserver client1 = loggedEmployees.remove(emp.getId());
        if(client1 == null)
            throw new MyException("Employee " + employee.getUsername() + "is not logged in!");
    }

    private void notifyEmployees(Game game){
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);
        for(IObserver client: loggedEmployees.values()){
            executor.execute(() ->{
                System.out.println("Game update in ServiceImpl" + game + " " + client);
                client.updateGame(game);
            });
        }
    }


    @Override
    public synchronized ArrayList<Game> findAllGames() {
        ArrayList<Game> result = new ArrayList<>();
        for(Game g : gamesRepo.findAll()){
            result.add(g);
        }
        return result;
    }
}

File: sem4/MPP/mpp-proiect-repository-mariaruncan/ProjCSharp/Server/ServicesImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Model;
using Persistence;
using Server.utils;
using Services;

namespace Server
{
    public class ServicesImpl : IServices
    {
        private EmployeesRepository employeesRepo;
        private GamesRepository gamesRepo;
        private TeamsRepository teamsRepo;
        private TicketPurchesesRepository ticketPurchesesRepo;
        private readonly IDictionary<long, IObserver> loggedEmployees;

        public ServicesImpl(EmployeesRepository employeesRepo, GamesRepository gamesRepo, TeamsRepository teamsRepo, 
            TicketPurchesesRepository ticketPurchesesRepo)
        {
            this.employeesRepo = employeesRepo;
            this.gamesRepo = gamesRepo;
            this.teamsRepo = teamsRepo;
            this.ticketPurchesesRepo = ticketPurchesesRepo;
            loggedEmployees = new Dictionary<long, IObserver>();
        }

        public List<Model.Game> FindAllGames()
        {
            return gamesRepo.findAll().ToList();
        }

        public void LogIn(string username, string pass, IObserver client)
        {
            Model.Employee emp = employeesRepo.FindByUsername(username);
            if(emp != null)
            {
                if(emp.HashedPassword != Hash.hash(pass))
                    throw new MyException("Authentication failed.");
                if (loggedEmployees.ContainsKey(emp.Id))
                    throw new MyException("Employee already logged in!");
                loggedEmployees.Add(emp.Id, client);
            }
            else
                throw new MyException("Authentication failed.");
        }

        public void LogOut(Model.Employee employee, IObserver client)
        {
            Model.Employee emp = employeesRepo.FindByUsername(employee.Username);
            IObserver client1 = loggedEmployees[emp.Id];
            if (client1 == null)
                throw new MyException("Employee " + employee.Username + " is not logged in!");
            loggedEmployees.Remove(emp.Id);
        }

        public void Purchase(long gameId, int noOfTickets, string clientName)
        {
            Model.Game game = gamesRepo.findById(gameId);
            float totalPrice = game.PricePerTicket * noOfTickets;
            Model.TicketPurchase ticketPurchase = new Model.TicketPurchase(game, noOfTickets, totalPrice, clientName);
            ticketPurchesesRepo.add(ticketPurchase);
            game.SoldTickets += ticketPurchase.NoOfTickets;
            gamesRepo.update(game, gameId);
            notifyEmployees(gamesRepo.findById(gameId));
        }

        private void notifyEmployees(Model.Game game)
        {
            Console.WriteLine("notify game updated");
            foreach(var elem in loggedEmployees)
            {
                Console.WriteLine("notify client {0}", elem.Key);
                Task.Run(() => elem.Value.UpdateGame(game));
            }
        }
    }
}
